### ISSUE TITLE
enr-udp-port does not default to port

### ISSUE THREAD

**Issue by MysticRyuujin** [sigp/lighthouse#1660]
## Description

When specifying the `--enr-address` flag, `--enr-tcp-port` defaults to `--port` when not set, however, `--enr-udp-port` does not. Meaning, if you specify `--enr-address` and don't specify `--enr-udp-port` the beacon chain process will not run.

I think it make sense to default to port here.

## Version
```
sigp/lighthouse:latest
DIGEST:sha256:ad23d728888e11ae1e348c69cae1a6f228af5286f803400ec60f0fa9c052d12a
```
## Present Behaviour

enr-udp-port does not default to port when not specified

## Expected Behaviour

enr-udp-port defaults to port when not specified

## Steps to resolve

set `--enr-udp-port=9000`

**Comment by AgeManning** [comment]
This behaviour is by design. 

There are complexities involved with discovery, which are not very well documented which makes this asymmetric behaviour between tcp and udp reasonable (in my opinion). 

Setting the `--enr-address` has a very limited use-case, its primarily used for boot-nodes so that you can construct an ENR that other's can connect to. It really shouldn't be used for a regular client. When creating boot-nodes, because ENR's are somewhat opaque, it has often been the case that user's have set `--enr-address` without setting the `--enr-udp-port` and end up having non-contactable ENR's when setting up boot-nodes. Because of this, we now require the `--enr-udp-port` flag to be set if you are setting the `--enr-address`. This forces you to manually create a contactable ENR for a bootnode. 

Now ENR's can have optional UDP and TCP ports. It can be the case a user wants to have no TCP and no UDP. So defaulting to these ports limits the optionality of having this configuration (this is why we do not default to the udp port). A boot-node does not require TCP, so a bootnode can run happily with a misconfigured TCP port but it cannot function without a UDP port (this is where the asymmetry comes in). 

Lighthouse has a `boot_node` CLI command which allows you set up such bootnodes without the TCP option. The TCP port indicates that you have libp2p running and are contactable via the standard libp2p protocols. If you run a beacon node (and therefore construct a listening TCP port) we default to the standard TCP (as we optimistically assume you want your node contacted via TCP on libp2p ports). If you don't explicitly specify an external TCP port, we default to the internal one with no loss to the functionality of the discovery bootnode you have constructed. 

The reason `--enr-address` should not be used for normal operation, is that if you are behind a NAT, your external ENR address will start off as `None` and remain in this state. This indicates you are not-contactable and should prevent your ENR from being included in the global DHT (as other nodes cannot contact you). If you are behind a NAT and you specify an `--enr-address` and an `--enr-udp-port` this will create an ENR that says you are contactable at these addresses (even though you are not (because of the NAT)). This will make it difficult for other nodes to filter you out of the DHT and will end up polluting it with non-contactable nodes. 
If you are not behind a NAT, the normal operation of the node will start with the external ip of the ENR as None. As it finds nodes and as those nodes can contact you, it will automatically update your ENR to your contactable address. This will then be inserted into the larger DHT correctly. 

So in short, you shouldn't use `--enr-address` unless you are setting up a bootnode. In which case you must specify your external udp port and optionally set your external tcp port. If you don't want to set your external tcp port, use the `boot_node` cli param which will not start a beacon-node or any libp2p service.

**Comment by MysticRyuujin** [comment]
Thank you for the detailed response...

So coming from Prysm and Geth, they both have flags to advertise your external IP such as `--nat` on Geth, `--p2p-host-ip` and `--p2p-host-dns` on Prysm...so I assumed that this was the equivalent.

But it sounds like it's not? Is there an alternative, or is it really just better to let it dynamically figure it out?

**Comment by AgeManning** [comment]
In my opinion, I think its best to let it dynamically figure it out (for most use cases). 

This way, we will not pollute the DHT with misconfigurations (there are currently many dead nodes in there, which slows discovery queries for all live nodes). 

For most standard use cases, the dynamic adjustment should handle everything fine. I'm not super familiar with exactly what the prysm flags do under the hood. The reason we still have these flags in lighthouse is for more advanced use cases. There can be cases where a network is set up to receive on one socket addr and send on another. In this case, the dynamic adjustment won't work as expected and manual intervention is required. The other use case I mentioned was setting up boot-nodes. 

We have `--listen-address` which allows you to select the interface to listen on (this might be similar to `--p2p-host-ip`). We support DNS in discovery via the `--enr-address` flag, but this removes dynamic ENR updates and assumes your DNS is contactable. 

Our Discovery tries its best to filter dead/misconfigured nodes from the DHT, but it there still appears to be many misconfigurations in the current medalla DHT. I think we'd have a better DHT if everyone just used dynamic ENR updates.

**Comment by AgeManning** [comment]
I'll have a chat with the Geth guys and double check these flags to ensure there isn't an edge case we are missing. 
Will respond on this issue with any useful results. 

Thanks for bringing it to our attention.

**Comment by AgeManning** [comment]
I realise I didn't answer your question. 

`--enr-address` does what you expect. It will advertise your external IP address. It should be coupled with `--enr-udp-port`, to make sure discovery works. 

For Lighthouse, you can set these if you know their values. However there is a chance you misconfigure it, and if you are contactable on these addresses, Lighthouse will automatically set the values for you. 

The one catch is the TCP port. If you have a port-forward, where your external TCP port is different from your internal one, discovery will not update this and you should use the `--enr-tcp-port` to advertise your external TCP port. The main reason is that discovery is UDP based and cannot detect your external TCP port. 

Lighthouse is adding UPnP support very soon, which (if supported by your router) will set up the port forwards automatically for you. 

If you are not contactable, you can still participate in the network, it just means others won't discover and connect to you. I think it's more important that we don't pollute the DHT, than optimistically allowing users to try and manually configure their settings to make them contactable (which isn't a requirement to participate, although it does strengthen the network).

**Comment by MysticRyuujin** [comment]
At the end of the day my objective is to provide a healthy, stable, externally reachable node to support the network.

I am behind a NAT, but I port forward TCP/UDP 9000 and I can be reached by a mostly static DNS record (it is a dynamic DNS record that auto-updates if it does change, but it hasn't changed in 2+ years lol)

So I assume that setting `--enr-address=validator.domain.com --enr-udp-port=9000 --enr-tcp-port=9000` is accurate, and while probably not required or recommended for most users (as dynamic discovery would pick it up eventually), it's a valid configuration?

**Comment by AgeManning** [comment]
This is correct. 

Dynamic discovery should set this up automatically for you in < 30 seconds. You can tell by the info log "Address Updated". 

If you for some reason get these settings wrong, you will be adding a non-contactable node to the DHT, which can linger for a while, degrading discovery.

**Comment by AgeManning** [comment]
The `--nat` in geth uses a variety of mechanisms to set up automatic port forwarding. 
We are introducing UPnP very soon which will have an essentially equivalent use. 

I'm closing this issue as everything seems to be as expected. Feel free to re-open if further issues arise.
